### ISSUE TITLE
ZStream#throttleShape does not work when chunks are larger than 'units'

### ISSUE THREAD

**Issue by svroonland** [zio/zio#4190]
See https://scastie.scala-lang.org/ZKXDglYpSvW97rSpgXfmsQ

**Comment by heaven-born** [comment]
Hi @svroonland .  if you wait for 99 seconds there will be enough tokens in a bucket to emit the whole chunk containing 1000 elements. `fromIterable(1 to 20)` emits in 1 sec because the bucket is full of tokens at the beginning. The idea behind such behaviour is optimisation I guess.

**Comment by svroonland** [comment]
@heaven-born That makes sense, I didn't think to wait that long. 

Still I think the most intuitive behavior for this example is to emit 10 elements every second. 

The way I understand, chunking is done for performance reasons and as such should be mostly an implementation detail. But there are places like this and in #4189 where it 'leaks' through to the user. I suppose the question is to what degree this is desired/necessary.

**Comment by iravid** [comment]
Yes, currently `ZStream#throttleShape` operates on whole chunks and does not make any effort to split them.

This can be solved by, e.g., using `Chunk#take` on the chunk to grab the largest portion that fits in the token bucket and emit that, while keeping the rest of the chunk in a Ref. 

Another (simpler) solution is to just stick a `BufferedPull` and operate on single elements.

**Comment by guersam** [comment]
This behavior is a bit surprising to me as well.

For example, my usecase is to throttle the elements from a paginated list to pipe them to a rate-limited external API like this:

```scala
Stream
  .paginateZIO(1) { i =>
    fetchPage(apiAuthHeader, i).map { page =>
      val next = if (page.isLast) None else Some(i + 1)
      page.subscriptions -> next
    }
  }
  .flatMap(Stream.fromIterable(_))
  .rechunk(1) // this one is necessary to make `throttleShape` work as expected
  .throttleShape(5, 1.second)(_.length)
```

I guess this kind usage is quite common. As @svroonland pointed out chunking seems to be an implementation detail and it's error prone to let users keep the chunk sizes in their mind.

**Comment by guersam** [comment]
One remaining question is about the relationship between the chunk-aware cost function and the remaining elements.

Would there any use case that can be affected by the implicit chunk splitting suggested by @iravid?

**Comment by guersam** [comment]
Naive chunk splitting won't work because the arbitrary cost function can be effectul and not even growing as the chunk size is growing.

**Comment by hearnadam** [comment]
Is this still an issue with 2.x?
